File: _designs_001/Binary Search Tree Iterator/BSTIterator.java
public class BSTIterator {
    Stack<TreeNode> s;

    public BSTIterator(TreeNode root) {
        s = new Stack<TreeNode>();
        for (; root != null; root = root.left)
            s.push(root);
    }

    public boolean hasNext() {
        return !s.empty();
    }

    public int next() {
        TreeNode p = s.peek();
        int result = p.val;
        s.pop();
        for (p = p.right; p != null; p = p.left)
            s.push(p);
        return result;
    }
}

File: _designs_001/Binary Search Tree Iterator/BSTIterator.rb
class BSTIterator
  def initialize(root)
    @s = []
    until root.nil?
      @s.push(root)
      root = root.left
    end
  end

  def hasNext
    !@s.empty?
  end

  def next
    p = @s.pop
    r = p.val
    p = p.right
    until p.nil?
      @s.push(p)
      p = p.left
    end
    r
  end
end